### ISSUE TITLE
suggestions for ml job workflow?

### ISSUE THREAD

**Issue by alando46** [pangeo-data/pangeo#523]
I'm trying to whiteboard out options for a pangeo-based system that will:
- allow users to do all of standard pangeo things
- users can take advantage of dask-ml to experiment with training models in-notebook,
- users can also submit a model, maybe à la kubeflow or pachyderm to run as a scaled up training job which:
---- runs independently of whatever jupyter kernel the user initiated it from
---- probably also uses dask-ml to parallelize whenever possible
---- ships training metrics to some sort of dashboard
---- handles model and training data version control
---- saves model weights somewhere

I'm mostly curious if anyone in the Pangeo community has tried implementing a deployment that allows for users to submit training jobs that run out-of-notebook. Alternatively, even if you haven't tried such a thing but maybe know of something already within the dask/pangeo ecosystem that would support such functionality, I'd be grateful to hear your thoughts :)

**Comment by jhamman** [comment]
Hi @alando46 - thanks for opening this issue. I think you could wire up all of this with existing technology in a Pangeo deployment. Do you have an example of a dashboard of ML metrics? I don't know of anything that fits this exact use case but I image you could do it. 

> runs independently of whatever jupyter kernel the user initiated it from

Users can currently run dask workers in kubernetes pods outside of the jupyter kernel (and pod). I'm not specifically aware of anyone running non-dask pods that look particularly different from the jupyter image but it would be possible.

**Comment by alando46** [comment]
Thanks for sharing your thoughts @jhamman. In terms of metrics thinking of something along the lines of Tensorboard or [Comet](https://www.comet.ml) (this seems to be an expensive enterprise version of the same idea as Tensorboard but with scikit-learn support and a lot more bells and whistles).

It sounds like there might be 2 ways to allow for a user to initiate a training "job" in notebook (probably via a client) that will then persist, run, and save its weights after the user has logged off.

1. Use the python client of kubectl to manually spin up another pod that will run the training job

2. Serialize the model in-notebook and post it to an api which would then run the training job

In terms of transferring training data, would it make sense/be possible to serialize the dask graph that represents the tasks to be done on data, and ship that to either the new kubernetes pod or training model api?

**Comment by jhamman** [comment]
@alando46 - I spent some time getting more familiar with kubeflow this weekend. It really seems like it could fit in with the Pangeo architecture/approach quite nicely. From my early understanding, you could probably do the following

- Add the dask-kubernetes capability to the kubeflow-jupyterhub chart to support distributed processing and training in an interactive environment
- Use tensorboard (and the related kubeflow apps) to provide a dashboard to users
- Have the option to use kubeflow to deploy models or to use kubeflow+dask-ml to deploy models.

I think each of these are just wiring exercises given the current infrastructure in the Pangeo/KubeFlow/Dask projects.

> In terms of transferring training data, would it make sense/be possible to serialize the dask graph that represents the tasks to be done on data, and ship that to either the new kubernetes pod or training model api?

I'm not sure if anyone has done this. Its a question worth asking to the dask-ml folks though.

**Comment by alando46** [comment]
@jhamman sorry, just saw this. I think I need to reconfigure my github notifications settings. I agree re potential utility of kubeflow + dask-kubernetes.

As a side note, In the last few weeks, I've been reading up on using Kubeflow to deploy models. One takeaway: Kubeflow relies on a project called [Seldon](https://github.com/SeldonIO/seldon-core/blob/master/readme.md) to actually serve the models. Seldon provides lots of great features but seems to be strongly configured to serve models via some sort of API endpoint.

If by deploying models we mean push a given docker container into some sort of batch process workflow (a use case that could be more common for geospatial datasets) [Argo workflows](https://argoproj.github.io/argo/) might be the best tool for the job. This should facilitate the ship-the-model/pipeline to the data instead of trying to stream raster files across the internet.

I believe that kubeflow's new pipelines functionality (as of v0.4) is actually built on Argo Workflows and uses a library called [Fairing](https://github.com/kubeflow/fairing) to create new Docker images to run model training jobs on-the-fly. Thus, I'm optimistically hoping that it once these models and their respective images are trained, that it shouldn't be too hard to wire something together that makes them easily swappable into custom Argo Workflows pipelines.

Apologies if this is more information than you care to hear but I spent a few weeks digging through this and figured it could be worth it to share here in case others are interested in prospective ML workflows. 

cc @nbren12

**Comment by nbren12** [comment]
Cool. thanks for that detective work. This just reminds me that there are just so many options in this pipelining/workflow space. Some from industry with a heavy focus on cloud/docker/kubernetes and a roughly equivalent set of tools from the academic world such as snakemake/nextflow/common-workflow. I've found this list is pretty comprehensive: https://github.com/pditommaso/awesome-pipeline.

**Comment by alando46** [comment]
Woah that is a comprehensive list. Thanks for sharing @nbren12

**Comment by alimanfoo** [comment]
Hi all, just wondering if anyone has got any further in investigating how to integrate a pangeo/jupyterhub deployment with a pipeline/workflow engine like snakemake/nextflow/airflow/...? We commonly use workflow engines for non-interactive large scale batch processing (e.g., of DNA sequence reads) and so it would be cool to have a jupyterhub deployment which could be the basis of either interactive analyses (scaling via dask-kubernetes) and non-interactive workflows (scaling via some other kubernetes-based mechanism). Some workflow engines already have kubernetes integration (e.g., [snakemake](https://snakemake.readthedocs.io/en/stable/executable.html#executing-a-snakemake-workflow-via-kubernetes)) so it wouldn't seem too crazy to imagine some kind of integration.

**Comment by nbren12** [comment]
Snakemake is one of my favorite tools, but it is very focused on file-based workflows. I wonder if their input/output model is flexible enough to work with zarr in the cloud for example.

**Comment by alimanfoo** [comment]
I have written snakemake workflows that build zarr on a local file system,
that works fine but usually I use the trick of touching a "success" file to
indicate that a task is complete, rather than trying to track multiple
files and/or directories. I haven't tried to use snakemake (or any workflow
engine) on the cloud yet, for any type of work, but if we could it would be
very convenient. We have migrated all our interactive work to our own
pangeo-like jupyterhub+dask system deployed on GCP via k8s, but for any
workflow work we currently drop back to using local HPC.

**Comment by nbren12** [comment]
There are more cloud-friendly tools like kubeflow, airflow and luigi. But IMO they are designed to work where each "task" is very large (e.g. a big PySpark job). On the other hand, Snakemake scales well to 1000s of intermediate size tasks, where each task is running a 10-second command line tool for example.

**Comment by TomAugspurger** [comment]
> either interactive analyses (scaling via dask-kubernetes) and non-interactive workflows (scaling via some other kubernetes-based mechanism)

I may be misunderstanding, but doesn't this require writing your "scale out" logic twice? Wouldn't it make sense to use a single tool for both case? (In other words, I don't see what prevents you from using Dask for the non-interactive workflow). You would still use a workflow tool like airflow to schedule the jobs, but use Dask to do the large-scale compute.

> But IMO they are designed to work where each "task" is very large (e.g. a big PySpark job).

Similar question here.

As an example, Anaconda processes all the [download logs](https://github.com/ContinuumIO/anaconda-package-data). That starts as an Airflow job, but uses Dask for parallelization / out-of-core.

**Comment by nbren12** [comment]
I think dask+airflow is a good solution. However, both tools represent computations as graphs and have schedulers, which seems redundant to me.

Bioinformatics has always used different workflow tools (e.g. snakemake) than industry, I think because they have many command line utilities for important pre-processing steps. These steps are typically small, but there are 1000s of them linked together in complicated ways. Make-like tools excel at orchestrating these many small jobs because the workflow is very similar to compiling a C code. Dask does the same but for python code. However, dask does not provide the on-disk persistence and resumability of workflow managers and my understanding is that Airflow’s scheduler does not scale to 1000s of steps very well.

Imagine a workflow where some very complicated pre-processing occurs which requires custom command line tools. Then, the outputs are dumped to zarr, some more computations performed with dask, and the outputs passed to yet another command line script to train an ML model. Probably, the best way with current tools would be to use snakemake for the pre-processing, dask for the python bit, and airflow for the orchestrating these “big” jobs together. Thus, we would use 3 tools, which each manage graphs of computations, but are each optimized to a specific niche. This doesn’t seem ideal.

**Comment by TomAugspurger** [comment]
> However, both tools represent computations as graphs and have schedulers, which seems redundant to me.

As you say, though, they're at very different levels. And they serve very different purposes. I don't see the redundancy.

That said, I'm not familiar with snakemake.

**Comment by cicdw** [comment]
Apologies for the plug here, but I came across this issue and wanted to point out that [Prefect](https://github.com/PrefectHQ/prefect) addresses many of the concerns raised here, and was built from the beginning with Dask workflows in mind.  We know of a few successful deployments running large scale machine learning-style workflows on Prefect + Dask, and we're always interested in collecting more use cases and feedback!

**Comment by nbren12** [comment]
Hey! Thanks for mentioning that. I saw Matt Rocklin's tweet about prefect and looked through your docs at some point. Is it possible resume interrupted or incomplete workflows?

**Comment by cicdw** [comment]
Ultimately yes it is.  If using the engine linked above, you can provide an incomplete dictionary of task states to control these sorts of things:
```python
flow.run(task_states={task_a: Success(result=10), ...})
```
The engine doesn't really provide any off-the-shelf persistence layer, so you'd have to hook something like that up if it's required for your workflow.

**Comment by nbren12** [comment]
Thanks that does seem like a workable interface. That said, I think having default on-disk persistence and automated tracking of completed tasks (e.g. by file system time-stamp) is what many scientists expect from a workflow tool.

**Comment by cicdw** [comment]
Yea for sure- our Cloud platform provides this but it isn’t a part of our open source tool.

**Comment by alimanfoo** [comment]
A few thoughts from the genomics/bioinformatics side, this isn't ML so
please tell me to raise a separate issue if that would be better.

For bioinformatics workflows there are three tools currently popular that
I'd like to use: snakemake, nextflow and cromwell. These are mostly used
for workflows that are stitching together many command line programs and/or
scripts.

On conventional HPC, task inputs and outputs will be files. On cloud,
inputs and outputs will be objects, but few command line programs can yet
read/write directly to object storage, so workflow tools handle copying
between object storage and some transient local file system.

I'd like to be able to log in to a jupyterhub portal and then have an
interface for launching and monitoring batch workflows via a tool like
snakemake, nextflow or cromwell.

One question is how to handle the relatively long-running nature of these
workflows. I.e., I need to have the workflow controller process and the
cluster running the workflow to keep running until the workflow finishes,
and not get killed as soon as I log out or my terminal becomes inactive.

I also need to be able to launch some kind of cluster which can run the
tasks. A kubernetes cluster is obvious, and at least snakemake and nextflow
can integrate with kubernetes IIUC. Ideally this would be as easy as it is
to launch a dask kubernetes cluster.

E.g., log in, launch a terminal, launch a cluster, run a snakemake
workflow, log out and go to lunch, come back later and log in to check
progress, etc.

Having some tracking of state and ability to resume workflows is critical.
Snakemake does this via file/object timestamps, I believe nextflow and
cromwell have some other way of storing state.

While integration with dask may make sense for airflow, I don't think it
would make sense for snakemake, nextflow or cromwell. These tools all have
their own logic for building dags and scheduling tasks, I don't think
there's anything to gain from layering on dask.

However, having some visual dashboard for monitoring workflow progress,
like the dask distributed dashboard, would be amazing.
